main/src/main/scala/sbt/Defaults.scala

```diff
@@ -717,8 +717,19 @@ object Defaults extends BuildCommon {
         val scalaProvider = appConfiguration.value.provider.scalaProvider
         val version = scalaVersion.value
         if (version == scalaProvider.version) // use the same class loader as the Scala classes used by sbt
-          Def.task(ScalaInstance(version, scalaProvider))
-        else
+          Def.task {
+            try {
+              val allJars = scalaProvider.jars
+              val Array(libraryJar) = allJars.filter(_.getName == "scala-library.jar")
```

Comment:
1. I don't think we can assume that the name `scala-library.jar` will remain that way for all implementations of Scala. That's why ScalaInstance returns `libraryJar`.
2. As of https://github.com/sbt/zinc/pull/677, it can return multiple files as `libraryJars`.

Reply:
Yeah. That is why I catch the exception and revert to the old approach if that doesn’t work (though I didn’t mean to leave the e.printStackTrace() in). I can look into incorporating the change from https://github.com/sbt/zinc/pull/677 though since I require a later version if zinc anyway.

Reply:
I figure though that any future changes to the scala compiler api require zinc/sbt coordinations anyway so I don’t know that it’s really possible to future proof this.

Reply:
I reworked this commit to use the new libraryJars api. The specific code block where you were worried about assuming that the scala library is called "scala-library.jar" seems completely safe the more I think about it. We only execute that code if the scalaVersion is the same as the application scala provider version. I can't figure out how that assumption would be broken without a lot of other things breaking. We also already assume in numerous places that the jars have speciifc names, see e.g. sbt.librarymanagement.ScalaArtifacts. I did, however, rework that code block to be more obviously safe in https://github.com/sbt/sbt/pull/4729/commits/d05344527ee670c6a07be28a3e48626235b9cab8.